### PR TITLE
Disable nuget.exe pack for project.json by default, add a fallback env var to enable it

### LINKED ISSUE #9149: [5.x] Make restore types that are project.json-specific obsolete
For example https://github.com/NuGet/NuGet.Client/blob/7bac015acc5b7e0161a058c8febc98abe2386d51/src/NuGet.Core/NuGet.LibraryModel/LibraryDependency.cs#L16 is not very useful anymore. 

Same applies to a lot of types throughout restore.

### LINKED ISSUE #11214: Disable nuget.exe pack for project.json by default, add a fallback env var to enable it
Related https://github.com/NuGet/Home/issues/7931

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/11214
Fixes: https://github.com/NuGet/Home/issues/9149

Regression? Last working version:

## Description

Disable nuget.exe pack for project.json by default, add a fallback env var to enable it

See: https://github.com/NuGet/Home/issues/7931 for more details.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [x] N/A 

- **Documentation**
  
  - [x] Documentation PR or issue filled - https://github.com/NuGet/docs.microsoft.com-nuget/issues/2524
  - **OR**
  - [ ] N/A

### CODE REVIEW

File: src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs [hunk 1]
```diff
@@ -713,10 +713,15 @@ public enum NuGetLogCode
         NU5040 = 5040,
 
         /// <summary>
-        /// Error_ReadmeCannotOpenFile
+        /// Error_ProjectJsonPack_Deprecated_And_Disabled
         /// </summary>
         NU5041 = 5041,
```
Comment (dominoFire) [inline src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs:718]: Are you changing this error semantics/meaning?

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs:718]: Oh no, I mistakenly pasted the description there.

I'll remove it. Nice catch.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs [hunk 2]
```diff
@@ -290,6 +300,18 @@ public Packaging.PackageBuilder CreateBuilder(string basePath, NuGetVersion vers
...
                 _usingJsonFile = true;
+
+                _ = bool.TryParse(_environmentVariableReader.GetEnvironmentVariable(NUGET_ENABLE_LEGACY_PROJECT_JSON_PACK), out bool enableLegacyProjectJsonPack);
+
+                if (!enableLegacyProjectJsonPack)
+                {
+                    Logger.Log(
+                        PackagingLogMessage.CreateError(
+                            string.Format(NuGetResources.Error_ProjectJson_Deprecated_And_Removed, builder.Id, NUGET_ENABLE_LEGACY_PROJECT_JSON_PACK),
+                            NuGetLogCode.NU5041));
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs]: I believe you wanted to use newly created NU5042 or NU5043 here.

File: src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs [hunk 3]
```diff
@@ -717,6 +717,16 @@ public enum NuGetLogCode
...
 
+        /// <summary>
+        /// Attempted to write files from multiple sources into the same location
+        /// </summary>
+        NU5042 = 5042,
```
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs]: Where is this one used here?

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs [hunk 4]
```diff
@@ -290,6 +300,18 @@ public Packaging.PackageBuilder CreateBuilder(string basePath, NuGetVersion vers
...
+                    return null;
```
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs:312]: Please include screenshot or text of this log from product.

File: src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs (hunk 3 above)
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs]: It's going to be NU5042. 

I hadn't rebased in a while, so I had to change the log code.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs (hunk 2 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs]: NU5042. Fixed.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs (hunk 4 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs:312]: Not sure what you mean you can refer to the resource. 

> `project.json` pack is disabled in the current NuGet version, and will be permanently removed in a future version.
Please consider migrating 'project1' to `PackageReference` and using the pack targets.
You can set the 'NUGET_ENABLE_LEGACY_PROJECT_JSON_PACK' environment variable to 'true' to temporarily reenable this functionality.

**Review by erdembayar — APPROVED** [review]

**Review by zivkan — APPROVED** [review]

**Review by jeffkl — APPROVED** [review]

**Comment by nkolev92** [comment]
Hey @aortiz-msft

Hoping I can get an approval for this.

This is part of the long ongoing effort to reduce technical debt in nuget.exe. 
project.json pack was a feature that never GA itself. 
In an earlier triage meeting we agreed to turn off this feature behind a feature flag.

This change affects `nuget.exe` only.

**Comment by nkolev92** [comment]
Ready to be merged, rejected for 6.0, waiting for dev to be 6.1 to merge.

**Pull request merged**